### PR TITLE
Tweaks to the property system

### PR DESCRIPTION
**Pull request by havocp**
This is all my changes in one PR, in case that is easier than viewing the previous two in isolation.
https://github.com/bokeh/bokeh/pull/3179 and https://github.com/bokeh/bokeh/pull/3182 could be closed if you'd rather see it all at once.

The end result:
 * add `foo = Override(default=whatever)`, so subclasses can change the default for a property without causing warnings.
 * split Property from PropertyDescriptor, so the object with get/set isn't the same as the object with type/default/help specification. I think this will make a lot of things easier, starting with `Override`.
 * replace 'changed' with 'non-default', dropping the concept of "reset"
 * allow some properties to skip serialization (https://github.com/bokeh/bokeh/pull/3179 )

More details in the individual commit messages.

`transform()` has kind of a weird signature in here because I was toying with using it to get rid of `to_dict`, but I don't really want to mix that into this PR or even try to clean up `to_dict` this week, I guess.
Short version, it's easy to clean up all the Units and `to_dict` special cases if `__get__` on a spec property can always return a dict, instead of returning the same string/number you used to set it. If we need to remember how you set the spec property and return in the original format, it's less clear how to clean up the special cases.

Anyway I think theming should be easier to do with some of these changes.

### CODE REVIEW

File: bokeh/properties.py [hunk 1]
```diff
@@ -332,7 +389,14 @@ def __init__(self, delegate, help="", use_prefix=True):
...
 class MetaHasProps(type):
-    def __new__(cls, class_name, bases, class_dict):
+    def __new__(meta_cls, class_name, bases, class_dict):
+        # HACK for now - we get called twice for some weird reason.
+        # __new__ used to be idempotent before PropertyDescriptor
+        # had to be transformed into Property, so it was OK, but now
+        # it isn't ... I think. Anyway, on the second call just do this?
+        # we need to figure this out for real and clean it up.
+        if '__properties__' in class_dict:
+            return type.__new__(meta_cls, class_name, bases, class_dict)
```
Comment (havocp) [inline bokeh/properties.py]: I have no idea what's up with this. One solution is to use only `__init__` and not `__new__`, possibly.
In any case it's not a new problem, the above hack is faster than the old code which just did everything twice (I think, at least).

**Comment by bryevdv** [comment]
Just for reference the reason behind giving back the value that was set and not a dict is so that things behave somewhat normally for users, especially in a notebook say. I'm not necessarily categorically against just returning the dict but that would need to be a wider discussion.

**Comment by havocp** [comment]
After taking a break from this and eating something I have some ideas on Units that could preserve the current behavior and also avoid special cases in the generic code. But not sure I should pile that into this PR. I may want to remove some params from transform() in this one though if they aren't immediately needed.

File: bokeh/charts/builder.py [hunk 2]
```diff
@@ -114,24 +113,31 @@ class Builder(HasProps):
...
+    The dimension labels that drive the position of the
+    glyphs. Subclasses should implement this so that the Builder
+    base class knows which dimensions it needs to operate on.
+    An example for a builder working with cartesian x and y
+    coordinates would be dimensions = ['x', 'y']. You should
+    then instantiate the x and y dimensions as attributes of the
+    subclass of builder using the :class:`Dimension
+    <bokeh.charts.properties.Dimension>` class. One for x, as x
+    = Dimension(...), and one as y = Dimension(...).
+    """
```
Comment (bryevdv) [inline bokeh/charts/builder.py]: hmm this probably won't play nice with our auto ref-doc generation tho that is just an FYI, not a high priority now

File: bokeh/models/map_plots.py [hunk 3]
```diff
@@ -83,10 +83,6 @@ def _check_missing_renderers(self):
     def _check_no_glyph_renderers(self):
         pass
 
-    map_options = Instance(GMapOptions, help="""
-    Options for displaying the plot.
-    """)
-
```
Comment (mattpap) [inline bokeh/models/map_plots.py]: I think this shouldn't have been removed. Note the type difference with the other case below.

File: bokeh/charts/builder.py (hunk 2 above)
Comment (havocp) [inline bokeh/charts/builder.py]: the docs on these are about how to make your own Builder subclasses, so hopefully not vital.

Comment (bryevdv) [inline bokeh/charts/builder.py]: Yah we can move them.

File: bokeh/properties.py [hunk 4]
```diff
@@ -421,9 +487,40 @@ def __new__(cls, class_name, bases, class_dict):
...
+        # Check for improperly overriding a Property attribute.
+        # Overriding makes no sense except through the Override
+        # class which can be used to tweak the default.
+        # Historically code also tried changing the Property's
+        # type or changing from Property to non-Property: these
+        # overrides are bad conceptually because the type of a
+        # read-write propery is invariant.
+        cls_attrs = cls.__dict__.keys() # we do NOT want inherited attrs here
+        for n in cls_attrs:
+            for b in bases:
```
Comment (mattpap) [inline bokeh/properties.py]: Would you mind using longer names? You write once, I read multiple times. There are cases where short names would be preferred (e.g. implementing an algorithm from a scientific paper, for 1-1 correspondence of names), but this is definitively not one of those.

File: bokeh/properties.py (hunk 1 above)
Comment (bryevdv) [inline bokeh/properties.py]: Can you elaborate on how to reproduce this? I was playing around in ipython and I could only see it getting called once per class definition.

File: bokeh/models/map_plots.py (hunk 3 above)
Comment (havocp) [inline bokeh/models/map_plots.py]: possibly we should remove the one in the base class instead? It's wrong to have it in both base and sub class with different types though, for sure.

File: bokeh/properties.py (hunk 1 above)
Comment (havocp) [inline bokeh/properties.py]: I was just doing `python -c import bokeh.model.Model` I think. also something in the pytest suite causes it. I believe with py3.

Comment (havocp) [inline bokeh/properties.py]: it's only for Model if that wasn't clear. Related to the extra Viewable metaclass maybe. Change this check to throw an exception the second time through and you should see it.

File: bokeh/properties.py (hunk 4 above)
Comment (havocp) [inline bokeh/properties.py]: agreed, will do

File: bokeh/properties.py (hunk 1 above)
Comment (bryevdv) [inline bokeh/properties.py]: Ah. So, one issue might be that one of the `__new__` uses `super` and the other calls `type.__new__`

Comment (bryevdv) [inline bokeh/properties.py]: No that's not it, this is the issue:
```
@add_metaclass(Viewable)
class Model(HasProps, CallbackManager):
```

`Model` inherits from `HasProps` so that triggers `MetaHasProps.__new__` but `Viewable` also inherits from `MetaHasProps` which will call `MetaHasProps.__new__`  via `super`

Comment (havocp) [inline bokeh/properties.py]: what confuses me is that both `__new__` seem to get the same class dict, but presumably they each return a different class. But both classes can't be used right, so one of them must just get tossed? Doesn't make sense... 

I wonder if it all works if we were using either the py2 or py3 native syntax and six is causing problems ?

Comment (bryevdv) [inline bokeh/properties.py]: OK, so I think everything works if I get rid of `add_metaclass` and just use the real python syntax, i.e., 
```
class HasProps(object, metaclass=MetaHasProps):
```
This might be the final nail in the coffin for any (accidental) python 2.6 support.

 Still verifying though

Comment (bryevdv) [inline bokeh/properties.py]: Wit the following diff, tests pass and an example works on python 3, and is supposed to work on python 2.
```
diff --git a/bokeh/model.py b/bokeh/model.py
index aafd94c..ea3462d 100644
--- a/bokeh/model.py
+++ b/bokeh/model.py
@@ -3,7 +3,9 @@ from __future__ import absolute_import, print_function
 import logging
 logger = logging.getLogger(__file__)
 
-from six import add_metaclass, iteritems
+from future.utils import with_metaclass
+
+from six import iteritems
 
 from .properties import Any, HasProps, List, MetaHasProps, String
 from .query import find
@@ -64,8 +66,7 @@ class Viewable(MetaHasProps):
         else:
             raise KeyError("View model name '%s' not found" % view_model_name)
 
-@add_metaclass(Viewable)
-class Model(HasProps, CallbackManager):
+class Model(with_metaclass(Viewable, HasProps, CallbackManager)):
     """ Base class for all plot-related objects """
 
     name = String()
diff --git a/bokeh/properties.py b/bokeh/properties.py
index 611180c..b425d68 100644
--- a/bokeh/properties.py
+++ b/bokeh/properties.py
@@ -41,21 +41,23 @@ to be understood by BokehJS.
 """
 from __future__ import absolute_import, print_function
 
+import logging
+logger = logging.getLogger(__name__)
+
 import re
 import types
 import difflib
 import datetime
 import dateutil.parser
+from future.utils import with_metaclass
 import collections
 from importlib import import_module
 from copy import copy
 from warnings import warn
 import inspect
-import logging
 import numbers
-logger = logging.getLogger(__name__)
 
-from six import string_types, add_metaclass, iteritems
+from six import string_types, iteritems
 
 from . import enums
 from .util.string import nice_join
@@ -396,7 +398,8 @@ class MetaHasProps(type):
         # it isn't ... I think. Anyway, on the second call just do this?
         # we need to figure this out for real and clean it up.
         if '__properties__' in class_dict:
-            return type.__new__(meta_cls, class_name, bases, class_dict)
+            import pdb; pdb.set_trace()
+            return super(MetaHasProps, meta_cls)(meta_cls, class_name, bases, class_dict)
         names = set()
         names_with_refs = set()
         container_names = set()
@@ -555,8 +558,7 @@ def abstract(cls):
 
     return cls
 
-@add_metaclass(MetaHasProps)
-class HasProps(object):
+class HasProps(with_metaclass(MetaHasProps, object)):
 
     def __init__(self, **properties):
         super(HasProps, self).__init__()
```
I guess this is a bug in `six.add_metaclass`. Metaclass inheritance is fairly obscure.

Comment (bryevdv) [inline bokeh/properties.py]: Obviously the `pdb` calls should be removed :) but the point is that they were never triggered

Comment (bryevdv) [inline bokeh/properties.py]: Confirmed this works on py3. I did not realize `future` was a third party lib. Maybe we can just copy the definition of `with_metaclass` into our own tree, it's very short:

http://python-future.org/_modules/future/utils.html

Here's a cleaned up diff
```
diff --git a/bokeh/model.py b/bokeh/model.py
index aafd94c..ea3462d 100644
--- a/bokeh/model.py
+++ b/bokeh/model.py
@@ -3,7 +3,9 @@ from __future__ import absolute_import, print_function
 import logging
 logger = logging.getLogger(__file__)
 
-from six import add_metaclass, iteritems
+from future.utils import with_metaclass
+
+from six import iteritems
 
 from .properties import Any, HasProps, List, MetaHasProps, String
 from .query import find
@@ -64,8 +66,7 @@ class Viewable(MetaHasProps):
         else:
             raise KeyError("View model name '%s' not found" % view_model_name)
 
-@add_metaclass(Viewable)
-class Model(HasProps, CallbackManager):
+class Model(with_metaclass(Viewable, HasProps, CallbackManager)):
     """ Base class for all plot-related objects """
 
     name = String()
diff --git a/bokeh/properties.py b/bokeh/properties.py
index 611180c..6a07ff3 100644
--- a/bokeh/properties.py
+++ b/bokeh/properties.py
@@ -41,21 +41,23 @@ to be understood by BokehJS.
 """
 from __future__ import absolute_import, print_function
 
-import re
-import types
-import difflib
+import logging
+logger = logging.getLogger(__name__)
+
+import collections
+from copy import copy
 import datetime
 import dateutil.parser
-import collections
+import difflib
+from future.utils import with_metaclass
 from importlib import import_module
-from copy import copy
-from warnings import warn
 import inspect
-import logging
 import numbers
-logger = logging.getLogger(__name__)
+import re
+import types
+from warnings import warn
 
-from six import string_types, add_metaclass, iteritems
+from six import string_types, iteritems
 
 from . import enums
 from .util.string import nice_join
@@ -390,13 +392,6 @@ _EXAMPLE_TEMPLATE = """
 
 class MetaHasProps(type):
     def __new__(meta_cls, class_name, bases, class_dict):
-        # HACK for now - we get called twice for some weird reason.
-        # __new__ used to be idempotent before PropertyDescriptor
-        # had to be transformed into Property, so it was OK, but now
-        # it isn't ... I think. Anyway, on the second call just do this?
-        # we need to figure this out for real and clean it up.
-        if '__properties__' in class_dict:
-            return type.__new__(meta_cls, class_name, bases, class_dict)
         names = set()
         names_with_refs = set()
         container_names = set()
@@ -487,7 +482,7 @@ class MetaHasProps(type):
             path = class_dict["__example__"]
             class_dict["__doc__"] += _EXAMPLE_TEMPLATE % dict(path=path)
 
-        return type.__new__(meta_cls, class_name, bases, class_dict)
+        return super(MetaHasProps, meta_cls).__new__(meta_cls, class_name, bases, class_dict)
 
     def __init__(cls, class_name, bases, nmspc):
         if class_name == 'HasProps':
@@ -555,8 +550,7 @@ def abstract(cls):
 
     return cls
 
-@add_metaclass(MetaHasProps)
-class HasProps(object):
+class HasProps(with_metaclass(MetaHasProps, object)):
 
     def __init__(self, **properties):
         super(HasProps, self).__init__()
```

Comment (bryevdv) [inline bokeh/properties.py]: Actually our use of `six` is very light too. Maybe this could be the start of a `bokeh.util.future` module that we eventually populate with the few things we need from `six` so that we can ditch it as a dependency.

**Comment by havocp** [comment]
Pushed another refinement, splitting PropertyGenerator from PropertyDescriptor. This lets us un-special-case both `Include` and `UnitsSpec._unit_type`, and I think it will also give us the pathway to clean up the `_units` properties because we can make custom subtypes of `Property` returned by `UnitsSpec.make_properties` and that should let us clean up the `to_dict` thing.

In light of this, `serialized` should probably move from `PropertyDescriptor` to `Property`, since one generator could produce both serialized and non-serialized properties.

File: bokeh/charts/builder.py (hunk 2 above)
Comment (rothnic) [inline bokeh/charts/builder.py]: The original intent with these properties that are overridden was to utilize the type checking, but I was wanting to avoid having each subclass implement __init__ and pass in the subclass-specific values that way, but didn't see a way to do this with the properties system. Since this wasn't an immediate priority, I left it as is, since it provided the documentation and modeled what the attributes should look like. I have a todo to go back and figure out a better way to do what I was trying to do. Since hasprops models instance properties, it was a misuse anyways.

**Comment by bryevdv** [comment]
Since works seems to be proceeding on this PR, I am going to close the other two.

**Comment by havocp** [comment]
Any other feedback here? I tried to address all comments so far. Also cleaned up the `to_dict` hack and added methods to property to avoid direct use of `Property.descriptor` which should make future custom `Property` subtypes easier to do.

File: bokeh/model.py [hunk 5]
```diff
@@ -264,16 +239,16 @@ def vm_serialize(self, changed_only=True):
 
         This method should be considered "private" or "protected",
         for use internal to Bokeh; use to_json() instead because
```
Comment (bryevdv) [inline bokeh/model.py]: Why not just go ahead and make it `_vm_serialize`?

File: bokeh/properties.py [hunk 6]
```diff
@@ -121,47 +123,71 @@ class _NotSet(object):
 class DeserializationError(Exception):
     pass
 
-class Property(object):
-    """ Base class for all type properties. """
+class PropertyGenerator(object):
+    """ Base class for objects that can generate Property instances. """
+    def __init__(self):
+        pass
```
Comment (bryevdv) [inline bokeh/properties.py]: don't need `__init__` here?

File: bokeh/model.py (hunk 5 above)
Comment (havocp) [inline bokeh/model.py]: Yeah - or get rid of it and put it right in to_json. I think the only complexity is there's some caller in models/ somewhere that I haven't really looked at. Maybe it can just use properties_with_values I'm not sure.

**Comment by bryevdv** [comment]
@havocp here is my quick survey of all app, glyph, and plotting examples:
```
app:
  - fourier_animated.py (not updating)
  - stocks/ (bar chart color change, selection overlay not showing)

glyphs:
  - data_tables.py (StringFormatter serialization issue)
  - data_tables_server.py (StringFormatter serialization issue)
  - glyph2_server.py (_pending_writes should be None )
  - line_animate.py (_pending_writes should be None)
  - linked_tap_server.py (_pending_writes should be None)
  - population_server.py (_pending_writes should be None)
  - prim_server.py (_pending_writes should be None)
  - taylor_server.py (_pending_writes should be None)
  - widgets_server.py (_pending_writes should be None)

plotting/file:
  - burtin.py (unrelated, resize issue)
  - image_url.py (unrelated I think)
  
plotting/notebook:
  - animated.ipynb
  - line_server.ipynb
  - random_walk.ipynb

plotting/server:
  - line_animate_widget.py (buttons have no effect)
  -selection_histogram.py (selections sometimes have no effect)

```

**Comment by bryevdv** [comment]
I think there are only a couple of issues actually possibly related to this PR above.

**Comment by bryevdv** [comment]
That "_pending_lock" is causing some failures on master now too, actually: https://travis-ci.org/bokeh/bokeh/jobs/93586378

**Comment by havocp** [comment]
The _pending_writes failure is not happening for me locally. I tried on master and on this PR branch, and I tried before and after upgrading to tornado 4.3. The travis output has the client-side stack trace which is useless; the server-side stack trace should show the problem clearly, but I can't seem to cause the error in order to see that...

**Comment by havocp** [comment]
Fixed the small stuff but help welcome on why I don't see those server errors. Maybe it will be obvious to me in the morning :-)

**Comment by bryevdv** [comment]
OK, maybe this is something to do with the environment? I re-ran this previously passing test, and now it is failing in the same way:

https://travis-ci.org/bokeh/bokeh/jobs/93438985

@damianavila any ideas on how to possibly diagnose what might be different?

**Comment by bryevdv** [comment]
I actually can't reproduce this locally either, even the ones I posted above earlier.

**Comment by bryevdv** [comment]
OK, I think I have something. When running the test suite, the examples start failing after the second example which is `data_tables_server.py`. This particular example has this error:
```
RuntimeError: Failed to push document: DeserializationError("Instance(CellFormatter) failed to deserialize reference to {'type': 'StringFormatter', 'id': 'c236ab59-8461-4d6b-a3e1-5f0bfd812668'}",)
```
If I disable this example in the tests, suddenly the subsequent examples start working. So I think there are two problem:

* something is preventing the serialization of the `StringFormatter` model
* this failure "leaks" into subsequent tests as a "pending write", presumably because we don't restart the server between examples, and we also re-use the default session

**Comment by bryevdv** [comment]
Now, I can't reproduce this locally on `master` or `bokeh_command` but they are failing on travis in a similar way. But I note that the property warnings that show up in the travis log are ones that I thought were introduced from this branch?

https://travis-ci.org/bokeh/bokeh/jobs/93622086#L372

Maybe this is more of the travis branch/PR merging magic I don't understand.

**Comment by bryevdv** [comment]
More info. If I run the server, the run data tables example I first get `StringFormatter` deserialization error. If I run again with the same running server, I then get the "_pending updates" problem. If I restart the server, it's back to `StringFormatter` deserialization error. So I definitely think the two bullet points above are on the right track. As to why `StringFormatter` won't deserialize, I am not sure, but I do see this:
```
IntEditor object at 0x1072fdf28>, '54848890-d39c-40f6-873c-3eca078274f0': <bokeh.models.widgets.tables.StringFormatter object at 0x107304208>}
(Pdb) models.get(json["id"])
(Pdb) json['id']
'b0ea53fc-259a-4304-9d44-3e6018faf0ad'
```
There is a `StringFormatter` in `models`, it just has a different id than the one in the ref. I checked and there is only one `StringFormatter` created in `data_tables_server.py` so I think this id mismatch is the proximate problem.

**Comment by bryevdv** [comment]
Keeping on... OK this problem persists even if I comment out all the cell formatters in `data_tables_server.py`! Checking, the default cell formatter is a `StringFormatter`. So I think there is some general problem serializing references properties in `TableColumns`. Maybe something to do with reference depth causing different id's to be generated at different times? Tables have columns have formatters. Does anything else go this many levels of references? I'm not sure what else the issue could be, but I hope this is helpful.

**Comment by bryevdv** [comment]
last note, I see this in both py2 and py3

**Comment by bryevdv** [comment]
It's the lambda initializer. 

Changing it to just
```
formatter = Instance(CellFormatter, StringFormatter(), help="""
    The cell formatter for this column. By default, a simple string
    formatter is used.
    """)

    editor = Instance(CellEditor, StringEditor(), help="""
    The cell editor for this column. By default, a simple string editor
    is used.
    """)
```
lets the example load. And also `python tests/travis/test server` successfully completes. 

Now, the widgets on data tables server example are not working (nothing updates, different problem, I think) but at least this narrows down the immediate problem hopefully.

**Comment by bryevdv** [comment]
Possibly the offending code path:
```
  /Users/bryan/work/bokeh/bokeh/model.py(221)collect_one()
-> cls._visit_immediate_value_references(obj, collect_one)
  /Users/bryan/work/bokeh/bokeh/model.py(189)_visit_immediate_value_references()
-> child = getattr(value, attr)
  /Users/bryan/work/bokeh/bokeh/properties.py(352)__get__()
-> return self._get(obj)
  /Users/bryan/work/bokeh/bokeh/properties.py(346)_get()
-> return self._get_default(obj)
  /Users/bryan/work/bokeh/bokeh/properties.py(366)_get_default()
-> default = self.class_default(obj.__class__)
  /Users/bryan/work/bokeh/bokeh/properties.py(327)class_default()
-> return self.descriptor.prepared_default(cls, self.name)
  /Users/bryan/work/bokeh/bokeh/properties.py(179)prepared_default()
-> default = self._raw_default()
  /Users/bryan/work/bokeh/bokeh/properties.py(171)_raw_default()
-> return self._copy_default(self._default)
  /Users/bryan/work/bokeh/bokeh/properties.py(166)_copy_default()
-> return default()
```
This seems to happen more than once. I think the issue is that default functions should only be called once per instance, after which the standard copy takes over. This seems to be borne out by the fact that the following diff:
```
diff --git a/bokeh/properties.py b/bokeh/properties.py
index eadbe83..533df3d 100644
--- a/bokeh/properties.py
+++ b/bokeh/properties.py
@@ -146,6 +146,7 @@ class PropertyDescriptor(PropertyGenerator):
         """ This is how the descriptor is created in the class declaration. """
         self._serialized = serialized
         self._default = default
+        self._created_default = None
         self.__doc__ = help
         self.alternatives = []
 
@@ -173,11 +174,12 @@ class PropertyDescriptor(PropertyGenerator):
     def prepared_default(self, cls, name):
         """The default transformed by prepare_value()."""
         overrides = cls._overridden_defaults()
-        if name in overrides:
-            default = self._copy_default(overrides[name])
-        else:
-            default = self._raw_default()
-        return self.prepare_value(cls, name, default)
+        if not self._created_default:
+            if name in overrides:
+                self._created_default = self._copy_default(overrides[name])
+            else:
+                self._created_default = self._raw_default()
+        return self.prepare_value(cls, name, self._created_default)
```
allows the data tables server example to run (again, modulo widget updates not working). I'm not sure if this is exactly the right fix or not.

**Comment by havocp** [comment]
Pushed a couple of fixes. data_table_server.py still has a problem which appears to be that coffeescript sends a Tooltip model to the server, and Python doesn't know anything about a Tooltip model - this is probably just a matter of adding the right property to the nonserializable properties list. In any case I don't think that Tooltip problem is caused by or related to this PR.

**Comment by havocp** [comment]
ha, was fixing the same thing :-) just saw your comment. I cached the default per-instance rather than per-property, so the function is run once per Model instance, but basically solved it in the same way.

**Comment by bryevdv** [comment]
OK!! Passing! Here is the list of server related example problems I found:
```

examples/app:

  - fourier_animated.py : draws initial plot, does not update/stream
  - stocks : charts colors wrong, missing selection overlay

examples/glyphs:

  - line_animate.py : stuttering/buffering just at startup

examples/embed:

  - animated.py : problem with autoload server (unexpected ',; token):

        load_libs(, function() {
        console.log("Bokeh: BokehJS plotting callback run at", new Date());
        Bokeh.embed.embed_items(docs_json, render_items, websocket_url);
 
examples/plotting/server:

  - line_animate.py : stuttering/buffering just at startup
  - line_animate_widget.py : not responsive to start/stop buttons
  - selection_histogram.py : some selections ignored/overwritten by previous
```
This is pretty short, and not all are related. I would like to merge this ASAP and fix these individually in other issues. Thoughts? ping @havocp @damianavila @mattpap @bokeh/core

**Comment by havocp** [comment]
It makes sense to me to merge this before it gets any bigger and then fix those examples. They are probably mostly unrelated, and some have fixes in the works or open issues already.

File: bokeh/charts/glyphs.py [hunk 7]
```diff
@@ -426,7 +433,7 @@ class BoxGlyph(AggregateGlyph):
     def __init__(self, label, values, outliers=True, **kwargs):
         width = kwargs.pop('width', None)
 
-        bar_color = kwargs.pop('color', None) or self.bar_color
+        bar_color = kwargs.pop('color', None) or kwargs.get('bar_color', None) or self.lookup('bar_color').default
```
Comment (havocp) [inline bokeh/charts/glyphs.py]: the `.default` here is broken, will fix

**Comment by bryevdv** [comment]
I think we also need to merge very soon because `master` is failing and new PRs off `master` are failing with the `"_pending" problem, and I think this would fix that?

**Comment by havocp** [comment]
yes, it's supposed to fix those failures

**Comment by bryevdv** [comment]
@bokeh/dev I am going to plan on merging this tomorrow morning CST unless there are objections.

**Comment by bryevdv** [comment]
@bokeh/dev planning to merge in ~1 hour if there are no objections.

**Comment by fpliger** [comment]
LGTM
